### ISSUE TITLE
Errors with v6.8.6 - Implementation of offline mode

### ISSUE THREAD

**Issue by SirUli** [ansible/ansible-lint#2651]
##### Summary

I do auto accept all minor releases of ansible-lint and hence i had a lot of CI/CD failures this morning after the v6.8.6 of ansible-lint has been installed which took [ansible-compat v2.2.4](https://github.com/ansible/ansible-compat/releases/tag/v2.2.4) with the new offline mode into play. I'm getting weird messages regarding the offline mode of ansible-lint:

```plaintext
WARNING  Collection install skipped because ansible versions before 2.14 do not support an offline mode.
```

A bit of fiddeling showed that it has partly something to do with the [syntax-check.py#96](https://github.com/ansible/ansible-lint/blob/main/src/ansiblelint/rules/syntax_check.py#L96) where the offline mode is explicitly configured. if i e.g. change this to offline=False, the errors described subsequently are gone.

What actually bothers me is that the installation of collections now has issues which are changing every time and are apparently related to the number of files that have to go through a syntax check and additionally change every time. See section "STEPS TO REPRODUCE" for details.

##### Issue Type

- Bug Report

##### Ansible and Ansible Lint details

```console (paste below)
ansible --version
ansible [core 2.13.5]
  config file = /home/users/SirUli/.ansible.cfg
  configured module search path = ['/home/users/SirUli/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/users/SirUli/.pyenv/versions/3.10.4/lib/python3.10/site-packages/ansible
  ansible collection location = /home/users/SirUli/.ansible/collections:/usr/share/ansible/collections
  executable location = /home/users/SirUli/.pyenv/versions/3.10.4/bin/ansible
  python version = 3.10.4 (main, Apr  5 2022, 14:24:25) [GCC 9.4.0]
  jinja version = 3.1.1
  libyaml = True

ansible-lint --version
ansible-lint 6.8.6 using ansible 2.13.5

```

- ansible installation method: pip
- ansible-lint installation method: pip

##### OS / ENVIRONMENT

RHEL7 / Ubuntu 20.04

##### STEPS TO REPRODUCE

Clone https://github.com/SirUli/ansiblelintv686

Then run ansible lint.

Examples:

```console
$ ansible-lint 
WARNING  Collection install skipped because ansible versions before 2.14 do not support an offline mode.
WARNING  Collection install skipped because ansible versions before 2.14 do not support an offline mode.

Passed with production profile: 0 failure(s), 0 warning(s) on 4 files.
```

No change, run again:

```console
$ ansible-lint 
WARNING  Collection install skipped because ansible versions before 2.14 do not support an offline mode.
WARNING  Collection install skipped because ansible versions before 2.14 do not support an offline mode.
WARNING  Retrying execution failure 250 of: ansible-galaxy collection install -v --force -p /home/users/SirUli/.cache/ansible-compat/d5494b/collections .
ERROR    Command returned 250 code:
Using /home/users/SirUli/.ansible.cfg as config file
Starting galaxy collection install process
Process install dependency map
Starting collection install process
Installing 'test.ansiblelintv686:1.25.1' to '/home/users/SirUli/.cache/ansible-compat/d5494b/collections/ansible_collections/test/ansiblelintv686'
to see the full traceback, use -vvv

ERROR! Unexpected Exception, this is probably a bug: [Errno 2] No such file or directory: '/home/users/SirUli/.cache/ansible-compat/d5494b/collections/ansible_collections/test/ansiblelintv686'

Command returned 250 code:
Using /home/users/SirUli/.ansible.cfg as config file
Starting galaxy collection install process
Process install dependency map
Starting collection install process
Installing 'test.ansiblelintv686:1.25.1' to '/home/users/SirUli/.cache/ansible-compat/d5494b/collections/ansible_collections/test/ansiblelintv686'
to see the full traceback, use -vvv

ERROR! Unexpected Exception, this is probably a bug: [Errno 2] No such file or directory: '/home/users/SirUli/.cache/ansible-compat/d5494b/collections/ansible_collections/test/ansiblelintv686'
```

No change, run again:

```console
$ ansible-lint 
WARNING  Collection install skipped because ansible versions before 2.14 do not support an offline mode.
WARNING  Collection install skipped because ansible versions before 2.14 do not support an offline mode.
WARNING  Retrying execution failure 250 of: ansible-galaxy collection install -v --force -p /home/users/SirUli/.cache/ansible-compat/d5494b/collections .
ERROR    Command returned 250 code:
Using /home/users/SirUli/.ansible.cfg as config file
Starting galaxy collection install process
Process install dependency map
Starting collection install process
Installing 'test.ansiblelintv686:1.25.1' to '/home/users/SirUli/.cache/ansible-compat/d5494b/collections/ansible_collections/test/ansiblelintv686'
to see the full traceback, use -vvv

ERROR! Unexpected Exception, this is probably a bug: [Errno 2] No such file or directory: 'test2.yml'

Command returned 250 code:
Using /home/users/SirUli/.ansible.cfg as config file
Starting galaxy collection install process
Process install dependency map
Starting collection install process
Installing 'test.ansiblelintv686:1.25.1' to '/home/users/SirUli/.cache/ansible-compat/d5494b/collections/ansible_collections/test/ansiblelintv686'
to see the full traceback, use -vvv

ERROR! Unexpected Exception, this is probably a bug: [Errno 2] No such file or directory: 'test2.yml'
```

##### Desired Behavior

ansible-lint continues working like in v6.8.5

##### Actual Behavior

See Steps to reproduce

**Comment by SirUli** [comment]
Found one more complaint about this: https://github.com/ansible/ansible-lint/issues/2645

And i spoke with other users where the error disappears when adding `--offline` but of course also prevents requirements installation. And they had to go back to 6.8.4 to make everything work again.

**Comment by nre-ableton** [comment]
I'm also having problems with `--offline` mode, as our CI setup requires a custom collections path to avoid hash conflicts under `~/.cache/ansible-compat`. However, even exporting `ANSIBLE_COLLECTIONS_PATH` does not work, I get this error:

```
WARNING  Retrying execution failure 1 of: ansible-galaxy collection install -v --offline -r requirements.yml -p /home/nre/.cache/ansible-compat/1a0050/collections
ERROR    No config file found; using defaults
Starting galaxy collection install process
Process install dependency map

ERROR    [WARNING]: The specified collections path '/home/nre/.cache/ansible-
compat/1a0050/collections' is not part of the configured Ansible collections
paths '/home/nre/Code/Ableton/ansible-role-nix/.ansible/collections'. The
installed collection will not be picked up in an Ansible run, unless within a
playbook-adjacent collections directory.
ERROR! Failed to resolve the requested dependencies map. Could not satisfy the following requirements:
* community.docker:3.2.1 (direct request)

Got 1 exit code while running: ansible-galaxy collection install -v --offline -r requirements.yml -p /home/nre/.cache/ansible-compat/1a0050/collections
```

Clearly from the error message, the collections path is not being respected.

**Comment by sfuerte** [comment]
> the collections path is not being respected

I'd say that "the offline mode is not being respected at all":

```sh
bash-5.1# ansible-lint --version
ansible-lint 6.8.6 using ansible 2.14.0

# WITHOUT the reqs YAML
bash-5.1# mv requirements.yml _test.yml
bash-5.1# ansible-lint  --offline ./__playbook__.yml ./packer-build.yml  tasks/*
WARNING  Overriding detected file kind 'yaml' with 'playbook' for given positional argument: ./__playbook__.yml
WARNING  Overriding detected file kind 'yaml' with 'playbook' for given positional argument: ./packer-build.yml

Passed with production profile: 0 failure(s), 0 warning(s) on 12 files.

# WITH the reqs YAML
bash-5.1# mv _test.yml requirements.yml
bash-5.1# ansible-lint  --offline ./__playbook__.yml ./packer-build.yml  tasks/*
WARNING  Role installation skipped because `ansible-galaxy role install` command does not support an offline mode.
WARNING  Retrying execution failure 1 of: ansible-galaxy collection install -v --offline -r requirements.yml -p <HOMEDIR>/.cache/ansible-compat/e2885f/collections
ERROR    Using /tmp/build/a6b784e6/repository/ansible.cfg as config file
Starting galaxy collection install process
Process install dependency map

ERROR    [WARNING]: The specified collections path '<HOMEDIR>/.cache/ansible-
compat/e2885f/collections' is not part of the configured Ansible collections
paths '<HOMEDIR>/.ansible/collections:/usr/share/ansible/collections'. The
installed collection will not be picked up in an Ansible run, unless within a
playbook-adjacent collections directory.
ERROR! Failed to resolve the requested dependencies map. Could not satisfy the following requirements:
* ansible.posix:* (direct request)

Got 1 exit code while running: ansible-galaxy collection install -v --offline -r requirements.yml -p <HOMEDIR>/.cache/ansible-compat/e2885f/collections
```

**Comment by ssbarnea** [comment]
@rgl I think that the error is very clear: you do not have `ansible.posix` collection installed and installation fails because there is no offline method for installing it.

If the collection would already be present, you wound not get an error.

Offline mode does not mean that you would be able to use the linter without having dependencies available. Is just mean "do not go to galaxy to grab them".

**Comment by rgl** [comment]
@ssbarnea, Ah, thank you for pushing me to look at that part again! I was actually installing the collections in a previous step. But for some odd reason, the collection installation failed, BUT the build proceeded at https://github.com/rgl/my-ubuntu-ansible-playbooks/actions/runs/3496326639/jobs/5854108995#step:3:633 ... this happened because my dockerfile RUN heredoc have a bug, they do not check the collection installation exit code. Which happened at that particular build.

But even after fixing this, it still faills at https://github.com/rgl/my-ubuntu-ansible-playbooks/actions/runs/3497565572/jobs/5856819737#step:4:156 with:

```
+ ansible-galaxy collection list

# /usr/share/ansible/collections/ansible_collections
Collection        Version
----------------- -------
ansible.posix     1.4.0  
ansible.utils     2.7.0  
community.docker  3.2.1  
community.general 6.0.1  
kubernetes.core   2.3.2  
WARNING  Retrying execution failure 1 of: ansible-galaxy collection install -v --offline -r requirements.yml -p /root/.cache/ansible-compat/244210/collections
ERROR    Using /project/ansible.cfg as config file
Starting galaxy collection install process
Process install dependency map

ERROR    [WARNING]: The specified collections path '/root/.cache/ansible-
compat/244210/collections' is not part of the configured Ansible collections
paths '/root/.ansible/collections:/usr/share/ansible/collections'. The
installed collection will not be picked up in an Ansible run, unless within a
playbook-adjacent collections directory.
ERROR! Failed to resolve the requested dependencies map. Could not satisfy the following requirements:
* ansible.posix:1.4.0 (direct request)

Got 1 exit code while running: ansible-galaxy collection install -v --offline -r requirements.yml -p /root/.cache/ansible-compat/244210/collections
```

If you notice, it seems the collections are not installed at the expected directory. Do you have any idea of might be going on?

**Comment by sfuerte** [comment]
> @rgl I think that the error is very clear: you do not have `ansible.posix` collection installed and installation fails because there is no offline method for installing it.

and that's ok, that used to work prior 6.8.6 and kinda expected now.
Moreover, if you check my comment above, by simply removing the reqs YAML, it passes the linting with no problems at all. 

The linting job is running inside a Docker container, and for simplicity and keeping its size small, don't want to install all roles/collections and download extra every single time. Just need the basic linting at that step.

**Comment by SirUli** [comment]
I wonder how the original issue got turned into the issue where people use ansible-lint in a way that i completely different. I'm NOT using "--offline" and i still get issues. Thankfully #2673 is a duplicate of my issue and hence i'll continue there. It would have been wonderful to not mix the topics here.
